pkg/syncer/cnsoperator/controller/cnsnodevmbatchattachment/cnsnodevmbatchattachment_helper.go

```diff
@@ -470,3 +504,245 @@ func getVolumesToDetachFromVM(ctx context.Context, client client.Client,
+	}
+	log.Debugf("Successfully patched PVC: %s with annotations %+v", pvc.Name, updatedpvc.Annotations)
+	return nil
+}
+
+// pvcHasUsedByAnnotaion goes through all annotations on the PVC to find out if the PVC is used by any VM or not.
+func pvcHasUsedByAnnotaion(ctx context.Context, pvc *v1.PersistentVolumeClaim) bool {
+	log := logger.GetLogger(ctx)
+
+	if pvc.Annotations == nil {
+		log.Infof("No annotation found on PVC %s", pvc.Name)
+		return false
+	}
+
+	for key := range pvc.Annotations {
+		if strings.HasPrefix(key, attachedVmPrefix) {
+			log.Infof("Annotation with prefix %s found on PVC %s", attachedVmPrefix, pvc.Name)
+			return true
+		}
+	}
+
+	log.Infof("PVC %s does not contain any annotations with prefix %s", pvc.Name, attachedVmPrefix)
+	return false
+}
+
+// addPvcFinalizer adds the given VM as an annotation for the given PVC and
+// adds CNS finalizer to the PVC.
+func addPvcFinalizer(ctx context.Context, client client.Client,
+	k8sClient kubernetes.Interface,
+	pvcName string, namespace string, vmInstanceUUID string) error {
+	log := logger.GetLogger(ctx)
+
+	// Acquire lock on PVC
+	namespacedVolumeName := namespace + "/" + pvcName
+	actual, _ := VolumeLock.LoadOrStore(namespacedVolumeName, &sync.Mutex{})
+	instanceLock, ok := actual.(*sync.Mutex)
+	if !ok {
+		return fmt.Errorf("failed to cast lock for PVC: %s", namespacedVolumeName)
+	}
+	instanceLock.Lock()
+	log.Infof("Acquired lock for PVC %s", namespacedVolumeName)
+	defer func() {
+		instanceLock.Unlock()
+		log.Infof("Released lock for PVC %s", namespacedVolumeName)
+	}()
+
+	// Get PVC object from informer cache
+	pvc, err := commonco.ContainerOrchestratorUtility.GetPvcObjectByName(ctx, pvcName, namespace)
+	if err != nil {
+		log.Errorf("failed to get PVC object for PVC %s. Err: %s", pvcName, err)
+		return err
+	}
+
+	// If it is a shared PVC, add annotation indicating that the PVC is being used by this VM.
+	if isSharedPvc(ctx, pvc.Spec.AccessModes) {
+		log.Infof("PVC %s is shared", pvc.Name)
+		err = addPvcAnnotation(ctx, k8sClient, vmInstanceUUID, pvc)
+		if err != nil {
+			log.Errorf("failed to add annotation %s to PVC %s in namespace %s for VM %s", cnsoperatortypes.CNSPvcFinalizer,
+				pvcName, namespace, vmInstanceUUID)
+			return err
+		}
+		log.Infof("Successfully updated annotation on PVC %s for VM %s", pvcName, vmInstanceUUID)
+
+		// Obtain the latest PVC again with updated annotations.
+		pvc, err = commonco.ContainerOrchestratorUtility.GetPvcObjectByName(ctx, pvcName, namespace)
+		if err != nil {
+			log.Errorf("failed to get PVC object for PVC %s. Err: %s", pvcName, err)
+			return err
+		}
+	}
+
+	// If finalizer already exists, there is nothing to be done.
+	if controllerutil.ContainsFinalizer(pvc, cnsoperatortypes.CNSPvcFinalizer) {
+		// Finalizer already present on PVC
+		log.Infof("Finalizer %s not present on PVC %s", cnsoperatortypes.CNSPvcFinalizer, pvcName)
+		return nil
+	}
+
+	return k8s.PatchFinalizers(ctx, client, pvc,
```

Comment:
Perform error checking here ?

Reply:
PatchFinalizers returns error here which will be caught in the calling function anyway.